File: system-tests/tia-maven/src/test/java/com/teamscale/tia/TiaMavenSystemTest.java
package com.teamscale.tia;

import com.teamscale.report.testwise.model.ETestExecutionResult;
import com.teamscale.report.testwise.model.TestwiseCoverageReport;
import com.teamscale.test.commons.SystemTestUtils;
import com.teamscale.test.commons.TeamscaleMockServer;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.IOException;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;

/**
 * Runs several Maven projects' Surefire tests that have the agent attached and one of our JUnit run listeners enabled.
 * Checks that this produces a correct coverage report.
 */
public class TiaMavenSystemTest {

	private static TeamscaleMockServer teamscaleMockServer = null;

	@BeforeEach
	public void startFakeTeamscaleServer() throws Exception {
		teamscaleMockServer = new TeamscaleMockServer(SystemTestUtils.TEAMSCALE_PORT)
				.withAuthentication("build", "6lJKEvNHeTxGPhMAi4D84DWqzoSFL1p4")
				.acceptingReportUploads()
				.withImpactedTests("bar/UnitTest/utBla()", "bar/UnitTest/utFoo()",
						"bar/IntegIT/itBla()", "bar/IntegIT/itFoo()");
	}

	@AfterEach
	public void stopFakeTeamscaleServer() {
		teamscaleMockServer.shutdown();
	}

	@Test
	public void testMavenTia() throws Exception {
		SystemTestUtils.runMavenTests("maven-project");

		assertThat(teamscaleMockServer.availableTests).extracting("partition").contains("MyPartition");

		assertThat(teamscaleMockServer.uploadedReports).hasSize(2);

		assertThat(teamscaleMockServer.impactedTestRepositories).containsExactly("myRepoId", "myRepoId");
		assertThat(teamscaleMockServer.uploadRepositories).containsExactly("myRepoId", "myRepoId");

		assertThat(teamscaleMockServer.impactedTestCommits.get(0)).matches("abcd1337, .*");
		assertThat(teamscaleMockServer.impactedTestCommits.get(1)).matches("abcd1337, .*");
		assertThat(teamscaleMockServer.uploadCommits.get(0)).matches("abcd1337, .*");
		assertThat(teamscaleMockServer.uploadCommits.get(1)).matches("abcd1337, .*");

		TestwiseCoverageReport unitTestReport = teamscaleMockServer.parseUploadedTestwiseCoverageReport(0);
		assertThat(unitTestReport.tests).hasSize(2);
		assertThat(unitTestReport.partial).isTrue();
		assertAll(() -> {
			assertThat(unitTestReport.tests).extracting(test -> test.uniformPath)
					.containsExactlyInAnyOrder("bar/UnitTest/utBla()", "bar/UnitTest/utFoo()");
			assertThat(unitTestReport.tests).extracting(test -> test.result)
					.containsExactlyInAnyOrder(ETestExecutionResult.PASSED, ETestExecutionResult.PASSED);
			assertThat(unitTestReport.tests).extracting(SystemTestUtils::getCoverage)
					.containsExactly("SUT.java:3,6-7", "SUT.java:3,10-11");
		});

		TestwiseCoverageReport integrationTestReport = teamscaleMockServer.parseUploadedTestwiseCoverageReport(1);
		assertThat(integrationTestReport.tests).hasSize(2);
		assertAll(() -> {
			assertThat(integrationTestReport.tests).extracting(test -> test.uniformPath)
					.containsExactlyInAnyOrder("bar/IntegIT/itBla()", "bar/IntegIT/itFoo()");
			assertThat(integrationTestReport.tests).extracting(test -> test.result)
					.containsExactlyInAnyOrder(ETestExecutionResult.PASSED, ETestExecutionResult.PASSED);
			assertThat(integrationTestReport.tests).extracting(SystemTestUtils::getCoverage)
					.containsExactly("SUT.java:3,6-7", "SUT.java:3,10-11");
		});
	}

	@Test
	public void testPreferBranchAndTimestampOverRevisionWhenProvidedManually() throws IOException {
		SystemTestUtils.runMavenTests("maven-project", "-DteamscaleRevision=abcd1337", "-DteamscaleTimestamp=master:HEAD");

		assertThat(teamscaleMockServer.impactedTestCommits.get(0)).matches("abcd1337, null");
		assertThat(teamscaleMockServer.impactedTestCommits.get(1)).matches("abcd1337, null");
		assertThat(teamscaleMockServer.uploadCommits.get(0)).matches("abcd1337, null");
		assertThat(teamscaleMockServer.uploadCommits.get(1)).matches("abcd1337, null");
	}

	@Test
	public void testBaselineRevisionIsPreferred() throws IOException {
		SystemTestUtils.runMavenTests("maven-project", "-DbaselineRevision=rev1", "-DbaselineCommit=master:1234");

		assertThat(teamscaleMockServer.baselines).containsExactly("rev1, null", "rev1, null");
	}

	@Test
	public void testBaselineCommitIsUsed() throws IOException {
		SystemTestUtils.runMavenTests("maven-project", "-DbaselineCommit=master:1234");

		assertThat(teamscaleMockServer.baselines).containsExactly("null, master:1234", "null, master:1234");
	}

}

File: system-tests/tia-maven/src/test/kotlin/com/teamscale/tia/TiaMavenSystemTest.kt
package com.teamscale.tia

import com.teamscale.client.JsonUtils
import com.teamscale.report.testwise.model.ETestExecutionResult
import com.teamscale.report.testwise.model.TestwiseCoverageReport
import com.teamscale.test.commons.SystemTestUtils
import com.teamscale.test.commons.SystemTestUtils.coverage
import com.teamscale.test.commons.SystemTestUtils.runMavenTests
import com.teamscale.test.commons.TeamscaleMockServer
import org.assertj.core.api.Assertions.assertThat
import org.conqat.lib.commons.filesystem.FileSystemUtils
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.io.File
import java.io.IOException
import java.nio.file.Paths

/**
 * Runs several Maven projects' Surefire tests that have the agent attached, and one of our JUnit run listeners enabled.
 * Checks that this produces a correct coverage report.
 */
class TiaMavenSystemTest {
	private lateinit var teamscaleMockServer: TeamscaleMockServer

	@BeforeEach
	@Throws(Exception::class)
	fun startFakeTeamscaleServer() {
		teamscaleMockServer = TeamscaleMockServer(SystemTestUtils.TEAMSCALE_PORT)
			.withAuthentication("build", "6lJKEvNHeTxGPhMAi4D84DWqzoSFL1p4")
			.acceptingReportUploads()
			.withImpactedTests(
				"org/example/UnitTest/utBla()",
				"org/example/UnitTest/utFoo()",
				"org/example/UnitB1Test/utBlub()",
				"org/example/UnitB1Test/utGoo()",
				"org/example/UnitB2Test/utBlub()",
				"org/example/UnitB2Test/utGoo()",
				"org/example/IntegrationIT/itBla()",
				"org/example/IntegrationIT/itFoo()",
				"org/example/IntegrationB1IT/itBlub()",
				"org/example/IntegrationB1IT/itGoo()",
				"org/example/IntegrationB2IT/itBlub()",
				"org/example/IntegrationB2IT/itGoo()"
			)
	}

	@AfterEach
	fun stopFakeTeamscaleServer() {
		teamscaleMockServer.shutdown()
	}

	@Test
	@Throws(Exception::class)
	fun testMavenTia() {
		runMavenTests("maven-project", "-Dtia")

		assertThat(teamscaleMockServer.allAvailableTests).extracting("partition")
			.containsOnly("Unit Tests", "Integration Tests")

		assertThat(teamscaleMockServer.getSessions()).hasSize(2)
		val unitTestSession = teamscaleMockServer.getSession("Unit Tests")
		val integrationTestSession = teamscaleMockServer.getSession("Integration Tests")

		assertThat(integrationTestSession.getCommit()).matches("abcd1337:myRepoId, .*")
		assertThat(teamscaleMockServer.impactedTestCommits[0]).matches("abcd1337:myRepoId, .*")
		assertThat(teamscaleMockServer.impactedTestCommits[1]).matches("abcd1337:myRepoId, .*")

		val unitTestReport = unitTestSession.onlyTestwiseCoverageReport
		assertThat(unitTestReport.partial).isTrue()
		checkExpectedUnitTestCoverage(unitTestReport)
		checkExpectedIntegrationTestCoverage(integrationTestSession.onlyTestwiseCoverageReport)
	}

	@Test
	@Throws(IOException::class)
	fun testPreferBranchAndTimestampOverRevisionWhenProvidedManually() {
		runMavenTests(
			"maven-project", "-DteamscaleRevision=abcd1337",
			"-DteamscaleTimestamp=master:HEAD", "-Dtia"
		)

		assertThat(teamscaleMockServer.impactedTestCommits[0]).matches("abcd1337:myRepoId, null")
		assertThat(teamscaleMockServer.impactedTestCommits[1]).matches("abcd1337:myRepoId, null")
		assertThat(teamscaleMockServer.getSession("Unit Tests").getCommit())
			.matches("abcd1337:myRepoId, null")
	}

	@Test
	@Throws(IOException::class)
	fun testBaselineRevisionIsPreferred() {
		runMavenTests(
			"maven-project", "-DbaselineRevision=rev1", "-DbaselineCommit=master:1234", "-Dtia"
		)

		assertThat(teamscaleMockServer.baselines).containsOnly("rev1, null")
	}

	@Test
	@Throws(IOException::class)
	fun testBaselineCommitIsUsed() {
		runMavenTests("maven-project", "-DbaselineCommit=master:1234", "-Dtia")

		assertThat(teamscaleMockServer.baselines).containsOnly("null, master:1234")
	}

	/**
	 * Starts a maven process with the reuseForks flag set to "false". Checks if the coverage can be converted to a
	 * testwise coverage report afterward.
	 */
	@Test
	@Throws(Exception::class)
	fun testMavenTiaWithoutReuseForks() {
		runMavenTests("maven-project", "-Dtia", "-DreuseForks=false")
		val workingDirectory = File("maven-project")
		val testwiseCoverage = File(
			Paths.get(workingDirectory.absolutePath, "coverage", "target", "tia", "reports", "testwise-coverage-1.json").toUri()
		)
		val testwiseCoverageReport = JsonUtils.deserialize<TestwiseCoverageReport>(
			FileSystemUtils.readFile(testwiseCoverage)
		)
		checkExpectedUnitTestCoverage(testwiseCoverageReport)
	}

	companion object {
		private fun checkExpectedUnitTestCoverage(testwiseCoverageReport: TestwiseCoverageReport?) {
			assertThat(testwiseCoverageReport).isNotNull()
			assertThat(testwiseCoverageReport!!.tests)
				.extracting<String> { it.uniformPath }
				.containsExactlyInAnyOrder(
					"org/example/UnitTest/utBla()",
					"org/example/UnitTest/utFoo()",
					"org/example/UnitB1Test/utBlub()",
					"org/example/UnitB1Test/utGoo()",
					"org/example/UnitB2Test/utBlub()",
					"org/example/UnitB2Test/utGoo()"
				)
			assertThat(testwiseCoverageReport.tests)
				.extracting<ETestExecutionResult> { it.result }
				.allMatch { it == ETestExecutionResult.PASSED }
			assertThat(testwiseCoverageReport.tests)
				.extracting<String> {it.coverage}.containsExactly(
					"SUTA.java:7,10-11;UnitTest.java:10-11",
					"SUTA.java:7,14-15;UnitTest.java:15-16",
					"SUTB1.java:7,14-15;UnitB1Test.java:10-11",
					"SUTB1.java:7,10-11;UnitB1Test.java:15-16",  // Ensure that sources from the project itself (SUT2), but also other dependent projects is considered (SUT1)
					"SUTB1.java:7,14-15;SUTB2.java:7,14-15;UnitB2Test.java:10-12",
					"SUTB1.java:7,10-11;SUTB2.java:7,10-11;UnitB2Test.java:16-18"
				)
		}

		private fun checkExpectedIntegrationTestCoverage(integrationTestReport: TestwiseCoverageReport) {
			assertThat(integrationTestReport.tests)
				.extracting<String> { it.uniformPath }
				.containsExactlyInAnyOrder(
					"org/example/IntegrationIT/itBla()",
					"org/example/IntegrationIT/itFoo()",
					"org/example/IntegrationB1IT/itBlub()",
					"org/example/IntegrationB1IT/itGoo()",
					"org/example/IntegrationB2IT/itBlub()",
					"org/example/IntegrationB2IT/itGoo()"
				)
			assertThat(integrationTestReport.tests)
				.extracting<ETestExecutionResult> { it.result }
				.allMatch { it == ETestExecutionResult.PASSED }
			assertThat(integrationTestReport.tests)
				.extracting<String> { it.coverage }
				.containsExactly(
					"IntegrationIT.java:10-11;SUTA.java:7,10-11",
					"IntegrationIT.java:15-16;SUTA.java:7,14-15",
					"IntegrationB1IT.java:10-11;SUTB1.java:7,14-15",
					"IntegrationB1IT.java:15-16;SUTB1.java:7,10-11",
					"IntegrationB2IT.java:10-11;SUTB2.java:7,14-15",
					"IntegrationB2IT.java:15-16;SUTB2.java:7,10-11"
				)
		}
	}
}
